### PR TITLE
change TokenBuffer parser getDecimalValue to avoid using lossy double parsing

### PR DESCRIPTION
**Pull request by pjfanning**
Double parsing can lead to rounding and if you ultimately want to get a BigDecimal, you need to avoid parsing as a double and the creating a BigDecimal from the double.

The current code might round inadvertently and be more expensive due to doing 2 conversions.

### CODE REVIEW

**Comment by cowtowncoder** [comment]
I agree with the idea of avoiding wrong conversions. However, this particular code path is only ever invoked on `String`-valued numbers which I think is only used for... actually I am not 100% where -- almost certainly not when buffering input.
I think it'd be only when writing into `TokenBuffer` using `JsonGenerator.writeNumber(String)`.

So I can merge this for that case, but I was wondering if you have found a specific problem case that this solves?

**Comment by pjfanning** [comment]
> I agree with the idea of avoiding wrong conversions. However, this particular code path is only ever invoked on `String`-valued numbers which I think is only used for... actually I am not 100% where -- almost certainly not when buffering input. I think it'd be only when writing into `TokenBuffer` using `JsonGenerator.writeNumber(String)`.
> 
> So I can merge this for that case, but I was wondering if you have found a specific problem case that this solves?

I haven't found a test case where this method causes rounding issues but there is one where the new lazy number stuff does. The Lazy number code is basically this method with wrappers to delay execution. So if I need to modify the lazy path, it seems like a should make equivalent changes to the eager path.

**Comment by pjfanning** [comment]
@cowtowncoder I managed to add some tests that exercise the new code. The new TokenBufferTest tests fail without my change to TokenBuffer.

File: src/test/java/com/fasterxml/jackson/databind/util/TokenBufferTest.java [hunk 1]
```diff
@@ -194,6 +194,32 @@ public void testNumberOverflowLong() throws IOException
         }
     }
 
+    public void testBigIntAsString() throws IOException
+    {
+        try (TokenBuffer buf = new TokenBuffer(null, false)) {
+            BigInteger big = BigInteger.valueOf(Long.MAX_VALUE).add(BigInteger.valueOf(1));
+            buf.writeNumber(big.toString());
+            try (JsonParser p = buf.asParser()) {
+                // NOTE: oddity of buffering, no inspection of "real" type if given String...
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/util/TokenBufferTest.java:203]: I think we can actually solve that in near future, by buffering number-as-Strings as deferred value that retains type information (at very least integer/fp distinction)

**Review by cowtowncoder — APPROVED** [review]
Ok, you convinced me. I'll merge this in.

**Comment by cowtowncoder** [comment]
Merged to `master` as well, tested that no breakage with dataformats (wrt changes here).

**Pull request merged**